src/GridFieldActionsMenu.php

```diff
@@ -0,0 +1,315 @@
+use SilverStripe\ORM\DataObject;
+use SilverStripe\ORM\ArrayList;
+use SilverStripe\Versioned\Versioned;
+use SilverStripe\View\ArrayData;
+use SilverStripe\View\SSViewer;
+
+/**
+ * Provides an encapsulated list of actions to do with the current record
+ */
+class GridFieldActionsMenu implements
+    GridField_ColumnProvider,
+    GridField_URLHandler
+{
+    /**
+     * Whether to include the first Root tab in the actions list
+     *
+     * @var boolean
+     */
+    protected $showFirstTab = true;
+
+    /**
+     * The action items to show in the menu
+     *
+     * Example structure:
+     *
+     * <code>
+     * // Actions
+     * [
+     *     // Group
+     *     [
+     *         // Action
+     *         [
+     *             'Title' => 'Content',
+     *             'Link' => 'hrefme',
+     *             'Type' => 'link', // or 'versioning'
+     *         ],
+     *         ...
+     *     ],
+     *     ...
+     * ]
+     * </code>
+     *
+     * @var array
+     */
+    protected $actions = [];
+
+    public function __construct($showFirstTab = true)
+    {
+        $this->setShowFirstTab($showFirstTab);
+    }
+
+    public function augmentColumns($gridField, &$columns)
+    {
+        if (!in_array('Actions', $columns)) {
+            $columns[] = 'Actions';
+        }
+    }
+
+    public function getColumnsHandled($gridField)
+    {
+        return ['Actions'];
+    }
+
+    /**
+     * Construct a list of dropdown menu actions to provide for the menu. This includes a list
+     * of the Root level tabs from the given record's FieldList, and some Versioned actions
+     * (publish, unpublish etc depending on the state of the record) if the record is versioned.
+     *
+     * @param GridField $gridField
+     * @param DataObject $record
+     * @return $this
+     */
+    protected function buildDefaultActions($gridField, $record)
+    {
+        GridFieldExtensions::include_requirements();
+
+        $linkCallback = function ($action = null, $hash = null) use ($gridField, $record) {
+            $link = Controller::join_links($gridField->Link('item'), $record->ID, $action);
+            // @TODO hack workaround: && false here because some JS in the CMS is rewriting
+            // a link with a hash in it to the page we're on _now_ #anchor, as opposed to
+            // e.g link/set/here#anchor
+            return $hash && false ? "$link#$hash" : $link;
+        };
+
+        $this->addRootTabActions($record, $linkCallback);
+        $this->addVersionedActions($record, $linkCallback);
+
+        return $this;
+    }
+
+    /**
+     * Add each of the "Root" tabs to the actions for this component
+     *
+     * We expect that a tabbed list of fields will always have a singular root.
+     *
+     * @param DataObject $record
+     * @param callable $linkCallback
+     */
+    protected function addRootTabActions(DataObject $record, callable $linkCallback)
+    {
+        $tabSet = $record->getCMSFields()->first();
```

Comment:
I don't think it's feasible to build the entire CMS fields for each record in order to auto-list tabs, in terms of performance impact here. Those could contain dozens of complex fields, conditional logic, record fetches. We need to make these views fast to make them useful, and this makes it harder going forward. 

How about switching this to explicit registration? So the GridField created for content blocks declares that every record has a group which happens deep link into tabs. The API for this component doesn't need to know anything about tabs. 

The component probably shouldn't even know about the difference between actions like "publish" and links like "settings". They're just actions that contain enough metadata (e.g. a deep link into a tab) to determine on the frontend how to handle them.

Reply:
Do the "tab actions" even need to be added by default? Maybe just make this a capability, and allow users to manually addActiontoGroup() in their GridField_Config setup. You could eliminate all of this complexity.

Doing too much "just in case" the user may want it often becomes a chore to remove and replace those items anyway.